zcash_client_sqlite/src/lib.rs

```diff
@@ -776,21 +869,31 @@ impl<P: consensus::Parameters> WalletWrite for WalletDb<rusqlite::Connection, P>
 
             // If we have some transparent outputs:
             if d_tx.tx().transparent_bundle().iter().any(|b| !b.vout.is_empty()) {
-                let nullifiers = wdb.get_sapling_nullifiers(NullifierQuery::All)?;
-                // If the transaction contains shielded spends from our wallet, we will store z->t
+                // If the transaction contains spends from our wallet, we will store z->t
                 // transactions we observe in the same way they would be stored by
                 // create_spend_to_address.
-                if let Some((account_id, _)) = nullifiers.iter().find(
+                let sapling_from_account = wdb.get_sapling_nullifiers(NullifierQuery::All)?.into_iter().find(
                     |(_, nf)|
-                        d_tx.tx().sapling_bundle().iter().flat_map(|b| b.shielded_spends().iter())
+                        d_tx.tx().sapling_bundle().into_iter().flat_map(|b| b.shielded_spends().iter())
                         .any(|input| nf == input.nullifier())
-                ) {
+                ).map(|(account_id, _)| account_id);
+
+                #[cfg(feature = "orchard")]
+                let orchard_from_account = wdb.get_orchard_nullifiers(NullifierQuery::All)?.into_iter().find(
+                    |(_, nf)|
+                        d_tx.tx().orchard_bundle().iter().flat_map(|b| b.actions().iter())
+                        .any(|input| nf == input.nullifier())
+                ).map(|(account_id, _)| account_id);
+                #[cfg(not(feature = "orchard"))]
+                let orchard_from_account = None;
+
+                if let Some(account_id) = sapling_from_account.or(orchard_from_account) {
```

Comment:
```suggestion
                if let Some(account_id) = orchard_from_account.or(sapling_from_account) {
```
Although we do not currently create transactions that shield from multiple accounts simultaneously in such a way that `sapling_from_account` would differ from `orchard_from_account`, this is possible and could be done by other wallets or a future version of a zcash_client_sqlite-based wallet. In that case, ZIP 316 would recommend to associate the transaction with the most preferred protocol [for the purpose of choosing an outgoing ovk](https://zips.z.cash/zip-0316#usage-of-outgoing-viewing-keys), when creating the transaction. So that is also the most consistent thing to do here.